### PR TITLE
[pkg/ottl] Fix escape sequences in strings

### PR DESCRIPTION
**Pull request by quentinmit**
**Description:** <Describe what has changed.>

Fix ottl parsing to properly handle escape sequences.
@TylerHelmuth

**Link to tracking Issue:**
#23238

**Testing:** <Describe what testing was performed and which tests were added.>
Added a new unit test with the example from the linked issue

**Documentation:** <Describe the documentation added.>
None (pure bugfix)

### CODE REVIEW

**Review by TylerHelmuth — COMMENTED** [review]
Thanks for finding this solution - I remember spending a while on that regex and was unable to come up with a suitable replacement.

Please add some new scenarios in `e2e` that put strings to the test.  Some scenarios I think would be useful:
- Using a string that is a single backslash
- Using a string that is 2 backslashes
- Using a string that is 3 backslashes
- Using a string that is 4 backslashes
- A test for the `a("\\", "b")` scenario.

plus any edge cases you can think of.

**Comment by quentinmit** [comment]
@TylerHelmuth Alright, I added several tests to `e2e_test.go` for this.

File: pkg/ottl/e2e/e2e_test.go [hunk 1]
```diff
@@ -518,6 +518,44 @@ func Test_e2e_converters(t *testing.T) {
 				tCtx.GetLogRecord().Attributes().PutStr("test", "pass")
 			},
 		},
+		{
+			statement: `set(attributes["test"], "\\")`,
```
[the remark this answers is not in the document]
Comment (quentinmit) [reply pkg/ottl/e2e/e2e_test.go:522]: The backticks are to make it a raw string in Go. So the resulting OTTL is

```
set(attributes["test"], "\\")
```

which represents a single backslash.

`"\"` is an incomplete string, so I can't use it in an e2e test. I suppose I could add it as a failure case in `lexer_test.go` if you want?

**Review by TylerHelmuth — APPROVED** [review]
I am happy that the new and existing test cases prove this regex works.  @evan-bradley please review

**Comment by TylerHelmuth** [comment]
@quentinmit please add a bug_fix changelog entry

File: pkg/ottl/grammar.go [hunk 2]
```diff
@@ -433,7 +433,7 @@ func buildLexer() *lexer.StatefulDefinition {
 		{Name: `Bytes`, Pattern: `0x[a-fA-F0-9]+`},
 		{Name: `Float`, Pattern: `[-+]?\d*\.\d+([eE][-+]?\d+)?`},
 		{Name: `Int`, Pattern: `[-+]?\d+`},
-		{Name: `String`, Pattern: `"(\\"|[^"])*"`},
+		{Name: `String`, Pattern: `"(\\.|[^\\"])*"`},
```
Comment (evan-bradley) [inline pkg/ottl/grammar.go:436]:
```suggestion
		{Name: `String`, Pattern: `"(\\.|[^"])*"`},
```
Do we need the other backslash excluded here? Do any tests fail if we remove it, or can you come up with any that do?

**Review by evan-bradley — COMMENTED** [review]
Thank you @quentinmit!

Can you update `ottlfuncs/README.md` to remove all references to #23238?

File: pkg/ottl/grammar.go (hunk 2 above)
Comment (quentinmit) [reply pkg/ottl/grammar.go:436]: You need backslash excluded or you will accept strings with mismatched backslashes. e.g. `"\"` would be incorrectly accepted with that change. I added a test to `lexer_test.go` to confirm that, and as expected it fails if you remove the backslash you referenced.

**Comment by quentinmit** [comment]
@evan-bradley @TylerHelmuth I added a changelog entry and updated the README. Thanks!

File: pkg/ottl/grammar.go (hunk 2 above)
Comment (evan-bradley) [reply pkg/ottl/grammar.go:436]: Got it, I missed that when I played around with it. Thanks!

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**